File: api/telegram.py
import json
import logging
import os
import sys
from http.server import BaseHTTPRequestHandler
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from bot import Config, DGacademyBot, configure_logging


configure_logging()
_bot = None


def get_bot():
    global _bot
    if _bot is None:
        config = Config()
        config.validate()
        _bot = DGacademyBot(config)
    return _bot


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.respond(200, {"ok": True, "service": "DGacademy Telegram webhook"})

    def do_POST(self):
        expected_secret = os.getenv("TELEGRAM_WEBHOOK_SECRET", "").strip()
        received_secret = self.headers.get("X-Telegram-Bot-Api-Secret-Token", "")
        if expected_secret and received_secret != expected_secret:
            self.respond(401, {"ok": False, "error": "Unauthorized"})
            return

        try:
            length = int(self.headers.get("Content-Length", "0"))
            payload = self.rfile.read(length).decode("utf-8")
            update = json.loads(payload or "{}")
        except Exception:
            logging.exception("Could not parse Telegram update")
            self.respond(400, {"ok": False, "error": "Invalid JSON"})
            return

        try:
            get_bot().handle_update(update)
        except Exception:
            logging.exception("Webhook handler failed")
            self.respond(200, {"ok": True, "warning": "Update accepted but processing failed"})
            return

        self.respond(200, {"ok": True})

    def respond(self, status, payload):
        data = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

File: api/telegram.ts
import { handleTelegramUpdate, type TelegramUpdate } from "../src/bot.js";

type VercelRequest = {
  method?: string;
  headers: Record<string, string | string[] | undefined>;
  body?: unknown;
};

type VercelResponse = {
  status(code: number): VercelResponse;
  json(payload: unknown): void;
};

export default async function handler(req: VercelRequest, res: VercelResponse): Promise<void> {
  if (req.method === "GET") {
    res.status(200).json({ ok: true, service: "DGacademy Telegram webhook" });
    return;
  }

  if (req.method !== "POST") {
    res.status(405).json({ ok: false, error: "Method not allowed" });
    return;
  }

  const expectedSecret = process.env.TELEGRAM_WEBHOOK_SECRET?.trim();
  const receivedSecret = readHeader(req.headers, "x-telegram-bot-api-secret-token");

  if (expectedSecret && receivedSecret !== expectedSecret) {
    res.status(401).json({ ok: false, error: "Unauthorized" });
    return;
  }

  try {
    await handleTelegramUpdate(req.body as TelegramUpdate);
    res.status(200).json({ ok: true });
  } catch (error) {
    console.error("Webhook handler failed:", error);
    res.status(200).json({ ok: true, warning: "Update accepted but processing failed" });
  }
}

function readHeader(headers: VercelRequest["headers"], name: string): string {
  const value = headers[name] ?? headers[name.toLowerCase()];
  return Array.isArray(value) ? value[0] ?? "" : value ?? "";
}
